File: content-solutions/graph-advanced/uf-first-cycle.py
def solution(n, bridges):
    parent = list(range(n))
    size = [1] * n

    def find(x):
        root = x
        while parent[root] != root:
            root = parent[root]
        while x != root:
            nxt = parent[x]
            parent[x] = root
            x = nxt
        return root

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra == rb:
            return False
        if size[ra] < size[rb]:
            ra, rb = rb, ra
        parent[rb] = ra
        size[ra] += size[rb]
        return True

    for i, (a, b) in enumerate(bridges, 1):
        if find(a) == find(b):
            return i
        union(a, b)
    return 0

File: content-solutions/graph-advanced/uf-first-cycle.java
import java.util.*;

class Solution {
    int[] parent, size;

    void init(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    int find(int x) {
        int root = x;
        while (parent[root] != root) root = parent[root];
        while (x != root) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    }

    boolean union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (size[ra] < size[rb]) {
            int t = ra;
            ra = rb;
            rb = t;
        }
        parent[rb] = ra;
        size[ra] += size[rb];
        return true;
    }

    public int solution(int n, int[][] bridges) {
        init(n);
        for (int i = 0; i < bridges.length; i++) {
            int a = bridges[i][0], b = bridges[i][1];
            if (find(a) == find(b)) return i + 1;
            union(a, b);
        }
        return 0;
    }
}
